instrumentation/spring/spring-integration-4.1/library/src/test/groovy/ComplexPropagationTest.groovy

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import static io.opentelemetry.instrumentation.test.utils.TraceUtils.runUnderTrace
+
+import io.opentelemetry.api.GlobalOpenTelemetry
+import io.opentelemetry.instrumentation.spring.integration.SpringIntegrationTracing
+import io.opentelemetry.instrumentation.test.LibraryInstrumentationSpecification
+import io.opentelemetry.sdk.trace.data.SpanData
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.stream.Collectors
+import org.springframework.boot.SpringApplication
+import org.springframework.boot.SpringBootConfiguration
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.context.ConfigurableApplicationContext
+import org.springframework.context.annotation.Bean
+import org.springframework.context.event.EventListener
+import org.springframework.integration.channel.DirectChannel
+import org.springframework.integration.channel.ExecutorChannel
+import org.springframework.integration.config.GlobalChannelInterceptor
+import org.springframework.messaging.Message
+import org.springframework.messaging.SubscribableChannel
+import org.springframework.messaging.support.ChannelInterceptor
+import org.springframework.messaging.support.MessageBuilder
+import spock.lang.Shared
+
+class ComplexPropagationTest extends LibraryInstrumentationSpecification {
+
+  @Shared
+  ConfigurableApplicationContext applicationContext
+
+  def setupSpec() {
+    def app = new SpringApplication(ExternalQueueConfig)
+    applicationContext = app.run()
+  }
+
+  def cleanupSpec() {
+    applicationContext?.close()
+  }
+
+  def "should propagate context through a custom message queue"() {
+    given:
+    def sendChannel = applicationContext.getBean("sendChannel", SubscribableChannel)
+    def receiveChannel = applicationContext.getBean("receiveChannel", SubscribableChannel)
+
+    def messageHandler = new CapturingMessageHandler()
+    receiveChannel.subscribe(messageHandler)
+
+    when:
+    runUnderTrace("parent") {
+      sendChannel.send(MessageBuilder.withPayload("test")
+        .setHeader("theAnswer", "42")
+        .build())
+    }
+
+    then:
+    messageHandler.join()
+
+    assertTraces(2) {
+      SpanData sendChannelSpan
+
+      trace(0, 2) {
+        sendChannelSpan = span(1)
+
+        span(0) {
+          name "parent"
+        }
+        span(1) {
+          name "application.sendChannel"
+          childOf span(0)
+          hasNoLinks()
+        }
+      }
+      trace(1, 2) {
+        span(0) {
+          name "application.receiveChannel"
+          hasNoParent()
+          hasLink sendChannelSpan
```

Comment:
Why is this a link and not a parent? E.g. in Kafka instrumentation we use context from the message as a parent. And I believe that current semantic conventions also dictate this.

Reply:
@anuraaga @trask What do you think about it?

Short version: right now the parent of the `MessageChannel` is always the "local" current span, and whatever comes with the message is added as a link. When there's no local span only link is added.
Another option is to always use the span context extracted from the message (and fall back to `Context.current()` if there's none).
Which one does make more sense as the default behavior? Is there any other option?

Reply:
We discussed this today during "special topics" meeting and it seems that using propagated context as parent makes more sense than using current or implicit context. When you use message listener (and in case of `SubscribableChannel` that is what you do, right), then your code will be called by the framework code. And you usually don't care about framework loops. So it still makes sense to have PRODUCER span as your parent. You have only "processing" part, not "receiving" one.

Reply:
👍

Reply:
Done - thanks for the discussion!